### ISSUE TITLE
6.1.0.rc1 no superclass method `fork' - fork_tracker.rb

### ISSUE THREAD

**Issue by baerjam** [rails/rails#40603]
Upgraded app from 6.0.2.2 to 6.1.0.rc1 and having issues loading the app in both development and staging/production environments. 

```
bundle/ruby/2.5.0/gems/activesupport-6.1.0.rc1/lib/active_support/fork_tracker.rb:8:in `fork': super: no superclass method `fork' for #<Puma::Cluster:0x000055eab1096b58> (NoMethodError)
	from bundle/ruby/2.5.0/gems/puma-4.3.5/lib/puma/cluster.rb:138:in `block in spawn_workers'
	from bundle/ruby/2.5.0/gems/puma-4.3.5/lib/puma/cluster.rb:134:in `times'
	from bundle/ruby/2.5.0/gems/puma-4.3.5/lib/puma/cluster.rb:134:in `spawn_workers'
	from bundle/ruby/2.5.0/gems/puma-4.3.5/lib/puma/cluster.rb:469:in `run'
	from bundle/ruby/2.5.0/gems/puma-4.3.5/lib/puma/launcher.rb:172:in `run'
	from bundle/ruby/2.5.0/gems/puma-4.3.5/lib/puma/cli.rb:80:in `run'
	from bundle/ruby/2.5.0/gems/puma-4.3.5/bin/puma:10:in `<top (required)>'
	from bundle/ruby/2.5.0/bin/puma:23:in `load'
	from bundle/ruby/2.5.0/bin/puma:23:in `<main>'

```
In development the same error comes from Spring loading the app. Was able to get around the error by moving spring machinery back to bin files as described in #40527 

Is there something else I am missing with this upgrade or is there a problem here?

### System configuration
**Rails version**:
6.1.0.rc1

**Ruby version**: 
ruby 2.5.1p57 (2018-03-29 revision 63029) [x86_64-linux]

**Comment by rafaelfranca** [comment]
Are you able to reproduce in a new application?

The fact that the puma object doesn't have a `fork` method is very strange to me. This could be an incompatibility with some gem.

**Comment by baerjam** [comment]
Yes. New app with ruby version: 2.5.1p57 (2018-03-29 revision 63029) [x86_64-linux]
Reports the same error but from Spring. 

/home/jbaer/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activesupport-6.1.0.rc1/lib/active_support/fork_tracker.rb:8:in `fork': super: no superclass method `fork' for #<Spring::Application:0x0000562b9997ade0> (NoMethodError)
	from /home/jbaer/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/spring-2.1.1/lib/spring/application.rb:180:in `serve'
	from /home/jbaer/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/spring-2.1.1/lib/spring/application.rb:145:in `block in run'
	from /home/jbaer/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/spring-2.1.1/lib/spring/application.rb:139:in `loop'
	from /home/jbaer/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/spring-2.1.1/lib/spring/application.rb:139:in `run'
	from /home/jbaer/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/spring-2.1.1/lib/spring/application/boot.rb:19:in `<top (required)>'
	from /home/jbaer/.rbenv/versions/2.5.1/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
	from /home/jbaer/.rbenv/versions/2.5.1/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
	from -e:1:in `<main>'


It looks like upgrading ruby to 2.6.6 might resolve this issue.

**Comment by rafaelfranca** [comment]
hmmm. So maybe `fork` is not present in all object on Ruby 2.5. We need to investigate that.

**Comment by baerjam** [comment]
Confirmed error is not present with ruby 2.6.6

**Comment by p8** [comment]
I could reproduce this error in a new rails app with:
* ruby 2.5.1 and 2.5.3
* rails 6.1.0.rc1
* puma 4.3.6 and 5.0.4
* `workers ENV.fetch("WEB_CONCURRENCY") { 2 }` set in puma.rb
* With and without spring

The error doesn't seem to occur when I use any of the following:
* ruby 2.5.8
* rails 6.0.3.4

I tried ruby 2.5.5 and 2.5.6 but could not get thos working locally.

**Comment by p8** [comment]
Possible cause: https://github.com/rails/rails/commit/f3260272f2b7e5fb8b41daeeeee28af7f9329ac7

**Comment by p8** [comment]
I can reproduce the error with the following code.
It fails on ruby 2.5.3 but works on ruby 2.5.4.
Skipping making :fork private fixes it on 2.5.3

```
class Cluster
  def start
    fork { puts "forked!" }
  end
end

module CoreExt
  def fork(*)
    super
  end
end

module CoreExtPrivate
  include CoreExt
  private :fork
end

::Object.prepend(CoreExtPrivate)
Cluster.new.start
```
